### PR TITLE
Remove Material Asset on Material change

### LINKED ISSUE #3560: Material changes are lost when Image Element is disabled
If any material changes are done to a Image element, those will be reverted if you disable the entity.

Example: https://playcanvas.com/editor/scene/1244945

This happens on, and have only been tested on chrome Version 94.0.4606.81.

Repo Steps:
* Create new Image Element.
* Assign a Material to the Element
* Write code that in runtime will clone the material.
* Disable and enable the Element.
* The Image Element will now have the initial material assigned to it, as seen in the screenshot bellow. 

image

### PR DESCRIPTION
**Pull request by MushAsterion**
Fixes #3560 

Set `ImageElement#materialAsset` to `null` on `material` change if new material does not match `materialAsset` to avoid overwrite from `ImageElement#onEnable`. Important to use setter instead of `_materialAsset` to remove listeners.

I confirm I have read the [contributing guidelines](https://github.com/playcanvas/engine/blob/master/.github/CONTRIBUTING.md) and signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

File: src/framework/components/element/image-element.js [hunk 1]
```diff
@@ -1034,13 +1043,15 @@ class ImageElement {
...
 
-            this._materialAsset = _id;
-            if (this._materialAsset) {
-                const asset = assets.get(this._materialAsset);
+            this._materialAsset = null;
+            if (_id) {
+                const asset = assets.get(_id);
                 if (!asset) {
                     this.material = null;
+                    this._materialAsset = _id;
```
Comment (querielo) [inline src/framework/components/element/image-element.js:1051]: `this._materialAsset = _id;` is in both branches of the `if`.

Can we replace only 1037 by `this._materialAsset = _id || null;`?

**Review by querielo — APPROVED** [review]

File: src/framework/components/element/image-element.js (hunk 1 above)
Comment (MushAsterion) [reply src/framework/components/element/image-element.js:1051]: The point is that if `this._materialAsset` is not `null` when we set `this.material` to `null`, `this.material` setter will call `this.materialAsset` to `null` which we don't want. So I had to make sure it's `null` before assigning to the right value after processing with `material` reset.

It's also not possible to change `this._material` directly as it won't reset the displayed material.

So yeah a bit upside-down but I couldn't find any better way to do it...

Comment (willeastcott) [reply src/framework/components/element/image-element.js:1053]: @querielo Any further comments here or is this particular conversation resolved?

Comment (MushAsterion) [reply src/framework/components/element/image-element.js:1053]: I moved to the other solution
```
this._materialAsset = null;
this.material = null;

this._materialAsset = _id;
```

So it's more clear that the `null` assignment is only for material `null`ing and reduces the number of changed lines, as it's only to reduces call from `material` setter and avoid a loop between the two.

In fact the bug was coming from material setter not changing the `_materialAsset` and the enable callback using the `_materialAsset` if any, overriding any existing `_material` if it was from a different ID/no ID. So the main point is to remove `_materialAsset` inside `material` setter which was then making a loop with `materialAsset` setter if `_materialAsset` was non-`null` or different, why the `null`.

**Review by willeastcott — APPROVED** [review]

**Review by slimbuck — APPROVED** [review]
Approving, sorry for the long wait @MushAsterion !!

**Comment by mvaligursky** [comment]
It looks good, merging in, thanks all!

File: src/framework/components/element/image-element.js [hunk 2]
```diff
@@ -989,6 +989,15 @@ class ImageElement {
         }
 
         this._material = value;
+
+        // Remove material asset if changed
+        if (this._materialAsset) {
+            const asset = this._system.app.assets.get(this._materialAsset);
+            if (!asset || asset.resource !== value) {
+                this.materialAsset = null;
```
Comment (yaustar) [inline src/framework/components/element/image-element.js:997]: Should this line be `this._materialAsset = null`? as calling the setter for `materialAsset` also sets the `_material` to null as well?

**Comment by willeastcott** [comment]
This PR seems to be causing a problem reported by a forum user:

https://forum.playcanvas.com/t/cannot-set-properties-of-alphatest-for-materials-even-on-playcanvas-examples/31387

They're cloning the material on an element and replacing the element's material with the clone. This internally sets materialAsset = null which in turn sets material to null. I'm wondering if we should back this out. Got any thoughts @MushAsterion?

**Comment by yaustar** [comment]
The fix in https://github.com/playcanvas/engine/pull/5109 has not been merged to a release yet @willeastcott

**Comment by mvaligursky** [comment]
You're right! With so many patch releases, I was sure this was out already.

**Pull request merged**